src/inference/dreamSample.ad.js

```diff
@@ -0,0 +1,138 @@
+// Coroutine of the forward pass in dream training. Samples from the guide
+// outside of mapData and from the target inside of mapData.
+// Records the traces and observations.
+
+'use strict';
+'use ad';
+
+var _ = require('underscore');
+var util = require('../util');
+var ad = require('../ad');
+var Trace = require('../trace');
+var guide = require('../guide');
+
+// TODO - should merge it back with forwardSample.js or keep it seperated?
+// There are similarities in the general structure but ultimately also
+// signficant differences and so maybe it would be better to keep these
+// as two simlpe seperated modules rather than one merged..
+module.exports = function(env) {
+
+  function DreamSample(s, k, a, wpplFn, options) {
+    this.opts = util.mergeDefaults(options, {
+      samples: 1,
+      verbose: true,
+      params: {}
+    });
+
+    if (this.opts.samples <= 0) {
+      throw 'Invalid samples amount.';
+    }
+
+    this.params = this.opts.params;
+    this.wpplFn = wpplFn;
+    this.s = s;
+    this.k = k;
+    this.a = a;
+
+    // true when inside mapData (i.e. local model)
+    // false otherwise (i.e. global model)
+    this.insideMapData = true;
+
+    this.records = [];
+    this.currRecord;
+
+    this.coroutine = env.coroutine;
+    env.coroutine = this;
+  }
+
+  DreamSample.prototype = {
+
+    // Run the program as per the number of samples we are required to collect.
+    // As part of each such an execution sample, we create a record that contains
+    // a completed trace and a list of observations.
+    run: function() {
+
+      return util.cpsLoop(
+          this.opts.samples,
+
+          // Loop body.
+          function(i, next) {
+
+            // Records initialization.
+            var trace = new Trace(this.wpplFn, this.s, this.k, this.a);
+            this.currRecord = {trace: trace, observations: []};
+
+            return this.wpplFn(_.clone(this.s), function(s, val) {
+
+              // Record completion.
+              this.currRecord.trace.complete(val);
+              this.records.push(this.currRecord);
+              return next();
+            }.bind(this), this.a);
+          }.bind(this),
+
+          // Continuation.
+          function() {
+            env.coroutine = this.coroutine;
+            return this.k(this.s, this.records);
+          }.bind(this));
+
+    },
+
+    // Samples from the guide outside mapData and from the target inside mapData.
+    // Add this choice to the recorded trace.
+    sample: function(s, k, a, dist, options) {
+      var distribution = this.insideMapData ?
+          dist : (options && options.guide) || guide.independent(dist, a, env);
+      var _val = distribution.sample();
+      var val = this.ad && dist.isContinuous ? ad.lift(_val) : _val;
```

Comment:
Yea, I wasn't completely sure about when do we need lifting and so I added that since I've seen it's used in sample calls of `smc`, `mhkernel`, `hmckernel` and `initialize` (i.e. `mcmc`)..I'll remove it now and will be glad to talk with you or Daniel more about the lifting later.